File: app/src/main/java/com/chooloo/www/callmanager/ui/widgets/ListItem.java
package com.chooloo.www.callmanager.ui.widgets;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.chooloo.www.callmanager.R;
import com.chooloo.www.callmanager.databinding.ListItemBinding;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ListItem extends ConstraintLayout {

    private ListItemBinding binding;

    public ListItem(Context context) {
        super(context);
        binding = ListItemBinding.inflate(LayoutInflater.from(context));
        
        setUp();
    }

    public ListItem(Context context, AttributeSet attrs) {
        super(context, attrs);
        LayoutInflater.from(context).inflate(R.layout.list_item, this, true);

        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.ListItem, 0, 0);
        setBigText(a.getString(R.styleable.ListItem_bigText));
        setSmallText(a.getString(R.styleable.ListItem_smallText));
        setHeaderText(a.getString(R.styleable.ListItem_header));
        setImageDrawable(a.getDrawable(R.styleable.ListItem_src));
        a.recycle();

        setUp();
    }

    private void setUp() {
        ButterKnife.bind(this);

        setClickable(true);
        setFocusable(true);
    }

    public void setBigText(@Nullable String text) {
        binding.listItemBigText.setText(text == null ? "" : text);
    }

    public void setSmallText(@Nullable String text) {
        binding.listItemSmallText.setText(text == null ? "" : text);
        binding.listItemSmallText.setVisibility(text == null ? GONE : VISIBLE);
    }

    public void setHeaderText(@Nullable String text) {
        binding.listItemHeader.listItemHeader.setText(text == null ? "" : text);
        showHeader(text != null);
    }

    public void setImageDrawable(@Nullable Drawable image) {
        if (image != null) {
            binding.listItemImage.setImageDrawable(image);
        }
    }

    public void setImageUri(@Nullable Uri image) {
        binding.listItemImage.setImageURI(image);
    }

    public void showHeader(boolean isShow) {
        binding.listItemHeader.listItemHeader.setVisibility(isShow ? VISIBLE : INVISIBLE);
    }
}

File: app/src/main/java/com/chooloo/www/callmanager/ui/widgets/ListItem.kt
package com.chooloo.www.callmanager.ui.widgets

import android.content.Context
import android.graphics.drawable.Drawable
import android.net.Uri
import android.util.AttributeSet
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.widget.LinearLayout
import androidx.annotation.ColorInt
import com.chooloo.www.callmanager.R
import com.chooloo.www.callmanager.databinding.ListItemBinding

class ListItem(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleRes: Int = 0
) : LinearLayout(context, attrs, defStyleRes) {
    private var _binding: ListItemBinding

    init {
        _binding = ListItemBinding.inflate(LayoutInflater.from(context), this, true)

        context.obtainStyledAttributes(attrs, R.styleable.ListItem, 0, 0).also {
            setBigText(it.getString(R.styleable.ListItem_bigText))
            setSmallText(it.getString(R.styleable.ListItem_smallText))
            setImageDrawable(it.getDrawable(R.styleable.ListItem_src))
            setHeaderText(it.getString(R.styleable.ListItem_header))
            showHeader(_binding.listItemHeaderText.text != null)
            it.recycle()
        }

        TypedValue().also {
            context.theme.resolveAttribute(android.R.attr.selectableItemBackground, it, true)
            setBackgroundResource(it.resourceId)
        }

        layoutParams = LayoutParams(MATCH_PARENT, WRAP_CONTENT)
        isClickable = false
        _binding.apply {
            listItemHeaderLayout.isClickable = false
            listItemHeaderLayout.isFocusable = false
            listItemPersonLayout.isClickable = true
            listItemPersonLayout.isFocusable = true
        }

    }

    fun setBigText(text: String?) {
        _binding.listItemBigText.text = text ?: ""
    }

    fun setSmallText(text: String?) {
        _binding.apply {
            listItemSmallText.text = text ?: ""
            listItemSmallText.visibility = if (text == null) GONE else VISIBLE
        }
    }

    fun setImageDrawable(image: Drawable?) {
        _binding.listItemImage.setImageDrawable(image)
    }

    fun setImageUri(image: Uri?) {
        _binding.listItemImage.setImageURI(image)
    }

    fun setImageBackgroundColor(@ColorInt color: Int) {
        _binding.listItemImage.setBackgroundColor(color)
    }

    fun setHeaderText(headerText: String?) {
        _binding.listItemHeaderText.text = headerText
    }

    fun showHeader(isShow: Boolean) {
        _binding.listItemHeaderLayout.visibility = if (isShow) VISIBLE else GONE
    }

    override fun setOnClickListener(onClickListener: OnClickListener?) {
        _binding.listItemPersonLayout.setOnClickListener(onClickListener)
    }

    override fun setOnLongClickListener(onLongClickListener: OnLongClickListener?) {
        _binding.listItemPersonLayout.setOnLongClickListener(onLongClickListener)
    }
}